wip-scripting.md

```diff
@@ -0,0 +1,320 @@
+<pre>
+  WIP: WIP-scripting
+  Layer: Consensus (hard fork)
+  Title: Add scripting to Witnet
+  Authors: Tomasz Polaczyk <tomasz@witnet.foundation>
+  Discussions-To: `#dev-general` channel on Witnet Community's Discord server
+  Status: Draft
+  Type: Standards Track
+  Created: 2022-08-09
+  License: BSD-2-Clause
+</pre>
+
+
+## Abstract
+
+Add scripts to value transfer transactions to allow spending UTXOs with complex conditions.
+
+## Motivation and rationale
+
+Support the following use cases:
+
+* Multisig
+* Atomic swap
+
+#### Multisig
+
+
+#### Atomic swap
+
+
+
+## Specification
+
+Scripting language, forth stack machine.
+
+Stack, dont mention alt stack because it is not used.
+
+A script input is an output pointer (UTXO) that needs a `redeem_script` and a `witness_script` to be spendable.
+
+A script output is the same as a normal output: the script address, value, and timelock.
+
+Script addresses are indistinguishable from normal addresses. A script address is calculated as the sha256 hash of the script bytes, truncated to 20 bytes. This is the same hash function used by normal addresses, which hash the public key instead of the script bytes.
+
+#### Supported transactions
+
+Script validation is described in the "Script execution" section. If the script validation is successful, the input can be spent. In case of error, the transaction is marked as invalid and cannot be included in the block.
+
+The following transaction types support scripts:
+
+* Value transfer
+
+Full support for script inputs and script outputs, this is the intented way to use scripts.
+
+When validating a `ValueTransferTransaction`, for each input, if that input has a `redeem_script`, do a script validation. Otherwise, if the input does not have a `redeem_script`, do a normal signature validation. Transactions can combine both script inputs and normal inputs. Script validation is described in the "Script execution" section.
```

Comment:
could it be possible for an input to have a `redeem_script` but no `witness_script` ?

Reply:
Yes, for example if the `redeem_script` always returns `true`. But since the `witness_script` contains the signatures this would make the input spendable by anyone, so it is basically a donation to the miner.